providers/azure/connection/azureinstancesnapshot/provider.go

```diff
@@ -0,0 +1,399 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package azureinstancesnapshot
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/cockroachdb/errors"
+	"github.com/rs/zerolog/log"
+	"go.mondoo.com/cnquery/v9/mrn"
+	"go.mondoo.com/cnquery/v9/providers-sdk/v1/inventory"
+	"go.mondoo.com/cnquery/v9/providers-sdk/v1/vault"
+	"go.mondoo.com/cnquery/v9/providers/azure/connection/auth"
+	"go.mondoo.com/cnquery/v9/providers/azure/connection/shared"
+	"go.mondoo.com/cnquery/v9/providers/os/connection"
+	"go.mondoo.com/cnquery/v9/providers/os/connection/snapshot"
+	"go.mondoo.com/cnquery/v9/providers/os/detector"
+	"go.mondoo.com/cnquery/v9/providers/os/id/azcompute"
+	"go.mondoo.com/cnquery/v9/providers/os/id/ids"
+)
+
+type scanTarget struct {
+	TargetType   string
+	InstanceName string
+	SnapshotName string
+}
+
+const (
+	SnapshotConnectionType shared.ConnectionType = "azure-snapshot"
+)
+
+type deviceInfo struct {
+	// the LUN number, e.g. 3
+	Lun string
+	// where the disk is mounted, e.g. /dev/sda
+	VolumePath string
+}
+
+// the instance from which we're performing the scan
+type azureScannerInstance struct {
+	instanceInfo
+}
+
+type mountInfo struct {
+	deviceName string
+	diskId     string
+	diskName   string
+}
+
+func (a *azureScannerInstance) getAvailableLun(mountedDevices []deviceInfo) (int32, error) {
+	takenLuns := []string{}
+	for _, d := range mountedDevices {
+		takenLuns = append(takenLuns, d.Lun)
+	}
+
+	availableLuns := []int32{}
+	// the available LUNs are 0-63, so we exclude everything thats in takenLuns
+	for i := int32(0); i < 64; i++ {
+		exists := false
+		for _, d := range takenLuns {
+			if d == fmt.Sprintf("%d", i) {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			availableLuns = append(availableLuns, i)
+		} else {
+			// log just for visibility
+			log.Debug().Int32("LUN", i).Msg("azure snapshot> LUN is taken, skipping")
+		}
+	}
+	if len(availableLuns) == 0 {
+		return 0, errors.New("no available LUNs to attach disk to")
+	}
+	return availableLuns[0], nil
+}
+
+// https://learn.microsoft.com/en-us/azure/virtual-machines/linux/azure-to-guest-disk-mapping
+// for more information. we want to find the LUNs of the data disks and their mount location
+func getMountedDevices(localConn *connection.LocalConnection) ([]deviceInfo, error) {
+	cmd, err := localConn.RunCommand("lsscsi --brief")
+	if err != nil {
+		return nil, err
+	}
+	if cmd.ExitStatus != 0 {
+		outErr, err := io.ReadAll(cmd.Stderr)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("failed to list logical unit numbers: %s", outErr)
+	}
+	// output looks like this:
+	// [0:0:0:0]    /dev/sda
+	// [1:0:0:0]    /dev/sdb
+	data, err := io.ReadAll(cmd.Stdout)
+	if err != nil {
+		return nil, err
```

Comment:
would be great to have a test around this parsing

Reply:
+1, i will add tests